File: FangZhiHu/app/src/main/java/com/tongqinyuan/fangzhihu/BaseActivity.java
package com.tongqinyuan.fangzhihu;

import android.os.Bundle;
import android.os.PersistableBundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import javax.inject.Inject;

import dagger.component.ActivityComponent;
import dagger.component.DaggerActivityComponent;
import qj.ccmtjf.com.lib.LoanApp;

/**
 * Created by tongqinyuan on 2018/3/8.
 */

public abstract class BaseActivity<P> extends AppCompatActivity {
    private static final String TAG = BaseActivity.class.getSimpleName();
    @Inject
    public P mPresenter;
    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(getLayoutId());
        injectActivity();

    }

    public void injectActivity() {
    }

    public ActivityComponent getActivityComponent() {
        ActivityComponent activityComponent = DaggerActivityComponent.builder()
                .appComponent(LoanApp.getInstance().getAppComponent()).build();
        Log.e(TAG,"==一个初始化="+activityComponent);
        return activityComponent;
    }

    abstract int getLayoutId();

}

File: FangZhiHu/KotlinOKhttp/src/main/java/com/kotlinokhttp/BaseActivity.kt
package com.kotlinokhttp

import android.os.Bundle
import android.os.PersistableBundle
import android.view.ContextMenu
import android.view.View
import com.kotlinokhttp.dagger.component.ActivityComponent
import com.kotlinokhttp.dagger.component.DaggerActivityComponent
import com.ytq.main.common.IBaseActivity
import com.ytq.main.common.IPresenter
import javax.inject.Inject

/**
 * Created by tongqinyuan on 2018/3/9.
 */
abstract class BaseActivity<P : IPresenter> :IBaseActivity() {

    /**
     * @jvmField 是让属性变成public static静态属性
     */
    @JvmField
    @Inject
    var mPresenter: P? = null

//    @Inject
//    lateinit var mPresenter : P//延迟初始化

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(getLayoutId())
        injectActivity()
        if (mPresenter != null) {
            mPresenter!!.AttachedView(this)
        }
    }


    fun getActivityComponent() : ActivityComponent{
        val activityComponent = DaggerActivityComponent.builder().run {
            this.appComponent(com.ytq.main.common.QianJiaApp.mInstance!!.getAppComponent())
            build()
        }
        return activityComponent

    }

    open fun injectActivity(){}

    abstract fun getLayoutId():Int


}